doc/source/ray-core/examples/rdt/grpo_contextual_bandits.py

```diff
@@ -0,0 +1,432 @@
+            dist = Categorical(logits=logits)
+            # Sample ACTION_DIM actions from each state's distribution.
+            acts = dist.sample((ACTION_DIM,))  # [ACTION_DIM, batch_size]
+            # Vectorized log_prob over the sampled actions for each state
+            logps = dist.log_prob(acts)  # [ACTION_DIM, batch_size]
+            acts = acts.transpose(0, 1).contiguous()  # [batch_size, ACTION_DIM]
+            logps = logps.transpose(0, 1).contiguous()  # [batch_size, ACTION_DIM]
+
+        # Create trajectory slices and enqueue trajectory slices for Scorer
+        slice_batch = {
+            "policy_version": self.policy_version,
+            "state": states,
+            "actions": acts,
+            "old_logps": logps,
+        }
+        self.scorer.enqueue_trajectory_batch.remote(slice_batch)
+
+    def update_weights(self, cuda_weights, version: int) -> bool:
+        """Apply GPU-to-GPU policy weight updates while pausing sampling."""
+        # Disable sampling while the GPU object transfer and weight load are in-flight.
+        with self._pause_generation():
+            first_tensor = next(iter(cuda_weights.values()))
+            if first_tensor.device.type != "cuda":
+                raise RuntimeError(
+                    "Expected CUDA tensors after GPU-to-GPU direct transfer"
+                )
+            self.model.load_state_dict(cuda_weights)
+            self.model.eval()
+            self.policy_version = int(version)
+        return True
+
+
+def run_once(total_steps: int) -> None:
+    """Run one end-to-end training session."""
+    replay_buf = ReplayBuffer.remote()
+    learner = Learner.remote(replay_buf)
+    scorer = Scorer.remote(replay_buf)
+    signal = SignalActor.remote()
+    generator = Generator.remote(scorer, signal)
+
+    # Initialize generator with current learner weights
+    ray.get(
+        generator.update_weights.remote(
+            learner.get_weights_ref.remote(), learner.get_version.remote()
+        )
+    )
+
+    num_steps = total_steps
+
+    # Pre-fill ReplayBuffer before starting PPO.
+    ray.get(generator.generate.remote(sample_unit_vector(batch_size=BATCH_SIZE)))
+
+    with tqdm(
+        total=num_steps, desc="Training", unit="step", leave=True, mininterval=1.0
+    ) as pbar:
+        for _ in range(num_steps):
+            states = sample_unit_vector(
+                batch_size=BATCH_SIZE
+            )  # [BATCH_SIZE, STATE_DIM]
+            generator.generate.remote(states)
+            step_result = ray.get(learner.step.remote())
+
+            pbar.set_postfix(
+                {
+                    "loss": f"{step_result['loss']:.3f}",
+                    "cosine_gap": f"{step_result['cosine_gap']:.3f}",
+                },
+                refresh=False,
+            )
+            pbar.update(1)
+
+            # Update generator with new weights/version
```

Comment:
Can you make sure sure all comments are full sentences?

I think running the linter tools should catch it too.

Reply:
Done